kong/plugins/pre-function/_handler.lua

```diff
@@ -11,37 +9,107 @@ return function(plugin_name, priority)
     VERSION = "0.3.1",
   }
 
-  function ServerlessFunction:access(config)
+  -- !!! Note thins function also executes fn_str !!!
+  -- If we support both old style (0.1.0) and new style (0.2.0+), this
+  -- early execution is mandatory, since it's our check for it.
+  local function load_function(fn_str)
+    local func1 = loadstring(fn_str)    -- load it
+    local _, func2 = pcall(func1)       -- run it
+    if type(func2) ~= "function" then
+      -- old style (0.1.0), without upvalues
+      return func1
+    else
+      -- this is a new function (0.2.0+), with upvalues
+      -- the first call to func1 above only initialized it, so run again
+      func2()
+      return func2
+    end
+  end
+
+  local function invoke(phase, config)
+    if not config then
+      return
+    end
+
+    -- old style phase check: phase is set, but is not this phase
+    -- new style phase check, no phase set, and no config entry for this phase
+
+    -- at this point, we are to assume we have returned early if phase does
+    -- not apply. So the following will always assume there's at least a
+    -- configuration for the phase, old and new style
+
+    local cache = config_cache[config] or {}
 
-    local functions = config_cache[config]
-
-    if not functions then
-      -- first call, go compile the functions
-      functions = {}
-      for _, fn_str in ipairs(config.functions) do
-        local func1 = loadstring(fn_str)    -- load it
-        local _, func2 = pcall(func1)       -- run it
-        if type(func2) ~= "function" then
-          -- old style (0.1.0), without upvalues
-          insert(functions, func1)
-        else
-          -- this is a new function (0.2.0+), with upvalues
-          insert(functions, func2)
-
-          -- the first call to func1 above only initialized it, so run again
-          func2()
+    -- old style phase support. config.functions apply to config.phase
+    if config.phase == phase then
```

Comment:
this is the "old-style", but this was never published. It was the initial version by @nijikokun .

So if we now add arrays per phase, then this "phase" indicator for the generic "functions" array can go imo.

Reply:
I thought about that and it's true... but then again, the interface meaning of `config.functions` kind of breaks apart: what phase do `config.functions` apply to?